### ISSUE TITLE
JSONExtract for Decimals doesn't work correct

### ISSUE THREAD

**Issue by Danisium** [ClickHouse/ClickHouse#69082]
When extracting a Decimal value from a JSON field in ClickHouse (24.7 and other) using JSONExtract, it appears that the function expects the Decimal value to be passed as a string within the JSON. However, JSON does not inherently require numbers to be represented as strings, and numeric values in JSON are commonly represented as numbers (not strings). This discrepancy can lead to precision loss when ClickHouse implicitly treats the JSON number as a float before converting it to Decimal.

Expected Behavior:

The direct extraction using JSONExtract should return the correct Decimal value with full precision, regardless of whether the JSON number is represented as a string or a numeric value.

Actual Behavior:
When extracting the Decimal value directly from a JSON number, the result shows precision issues, suggesting that the value was implicitly treated as a float, resulting in 111244542.0030000120206447 instead of 111244542.0030000000000000.
To achieve correct precision, the JSON number needs to be represented as a string and then cast to Decimal, which is not an inherent requirement of the JSON format.

Cause:
The issue seems to stem from the fact that ClickHouse expects Decimal values in JSON to be passed as strings, possibly to avoid precision loss during internal conversion. This is not aligned with the JSON specification, where numbers can be represented as numeric values and should not require string representation to maintain precision.

Workaround:
As a workaround, extract the value as a string using JSONExtractString and then cast it to Decimal:

SELECT JSONExtractString(dec, 'dec')::Decimal(32,16) FROM test;

Impact:
This behavior can lead to incorrect data interpretation and loss of precision, especially in financial or scientific applications where exact decimal values are critical.

Suggested Fix:
ClickHouse should correctly handle Decimal extraction from JSON numbers without requiring the values to be passed as strings, ensuring full precision is maintained.

Additional Information:
ClickHouse version:  24.8 or anyLast
https://fiddle.clickhouse.com/27debbf7-4bab-49e7-b9d5-38ab097905e7
create table test (
dec String    
)
Engine = Memory;

INSERT INTO test VALUES (1, now64(3), 'First', 1, '{"dec":111244542.003}');
INSERT INTO test VALUES (1, now64(3), 'First', 1, '{"dec":"111244542.003"}');

select JSONExtract(dec,'dec','Decimal(31,16)'),  JSONExtractString(dec,'dec')::Decimal(32,16) from test ;

**Comment by anselmodadams** [comment]
This happens since `Decimal` support was added to JSONExtract back in 21.8: https://fiddle.clickhouse.com/464b5250-302c-408e-912c-0018747ace9d

**Comment by MalKeshar** [comment]
> This happens since `Decimal` support was added to JSONExtract back in 21.8: https://fiddle.clickhouse.com/464b5250-302c-408e-912c-0018747ace9d, and can overflow for Decimal(32,16) on older versions when json attribute is not a string

No, there is real issue here even with the latest version of clickhouse
https://fiddle.clickhouse.com/9b790107-6681-4eb6-8233-a4d986f52041

```
WITH '{"amount":16.4}' AS json
SELECT
    JSONExtract(json, 'amount', 'Decimal64(6)'),
    CAST(JSONExtractString(json, 'amount'), 'Decimal64(6)')
 
   ┌─JSONExtract(json, 'amount', 'Decimal64(6)')─┬─CAST(JSONExtractString(json, 'amount'), 'Decimal64(6)')─┐
1. │                                   16.399999 │                                                    16.4 │
   └─────────────────────────────────────────────┴─────────────────────────────────────────────────────────┘
```
